### ISSUE TITLE
Add Merge method

### ISSUE THREAD

**Issue by akamensky** [go-git/go-git#942]
Somehow this project has `fetch` and `pull`, but missing `merge`?

FWIW, `pull` is a `fetch` immediately followed by `merge`, so how was the `pull` implemented  in the first place?

**Comment by fcharlie** [comment]
Implementing the merge feature in go-git is not an easy task; currently, there is no suitable implementation of the diff3 algorithm in Golang.

The few diff3 implementations that exist each have their own problems: [nasdf/diff3](https://github.com/nasdf/diff3) relies on github.com/sergi/go-diff just like go-git, but unfortunately, [sergi/go-diff](https://github.com/sergi/go-diff) is currently erroneous. Meanwhile, [epiclabs-io/diff3](https://github.com/epiclabs-io/diff3) is a reimplementation of [node-diff3](https://github.com/bhousel/node-diff3), based on the LCS algorithm; however, it is regrettable that the LGPL 3.0 licensing agreement may not be a good choice.

The maintainer of libgit2, Edward Thomson, has separated the diff3 implementation of git into the [libgit2/xdiff](https://github.com/libgit2/xdiff) project. This implementation has better compatibility with git and can be used with cgo or be rewritten in Golang, though this is also a significant undertaking.

In summary, implementing the merge functionality in go-git still presents some challenges; beyond the diff3 algorithm, there are many details that need to be addressed.

CC: @pjbgf

**Comment by gmax79** [comment]
Please make it possible to create merge commit without diff3 functionality. Leave the task of filling the commit to the program code. Just need to make a method that creates a new commit based on 2 parent ones and function that will be called for the merge files

**Comment by enisdenjo** [comment]
[epiclabs-io/diff3 is now MIT](https://github.com/epiclabs-io/diff3/pull/2)

**Comment by thegrumpylion** [comment]
Here's an example using plumbing objects and [epiclabs-io/diff3](https://github.com/epiclabs-io/diff3)
https://gist.github.com/thegrumpylion/15f20b9556ea7d75834d754687275f30

**Comment by caffeinated92** [comment]
Hi, My I know the status of this enhancement?

**Comment by fcharlie** [comment]
We recently open-sourced [HugeSCM](https://github.com/antgroup/hugescm), a cloud-native next-generation version control system. As part of this project, we implemented a complete three-way merge in pure Go — including both text-level (diff3) and tree-level merge — without shelling out to `git merge-file`.

We'd like to share this as a reference implementation that may be useful for go-git's merge support.

**Key components:**

- [`modules/diferenco`](https://github.com/antgroup/hugescm/tree/master/modules/diferenco) — Standalone diff3 merge library (reusable, no external dependencies)
- [`pkg/zeta/odb/merge.go`](https://github.com/antgroup/hugescm/blob/master/pkg/zeta/odb/merge.go) — Tree-level three-way merge orchestration

**Features:**

- Pure Go diff3 implementation with multiple diff algorithms (Histogram, Myers, ONP, Patience, Minimal, SuffixArray)
- Multiple conflict marker styles: merge (default), diff3, zdiff3
- Charset-aware merging (auto-detect and transcode GBK, Shift-JIS, etc.)
- Tree-level merge with rename detection, file/directory conflicts, mode conflicts
- Binary file handling via content sniffing + size threshold
- Pluggable merge drivers and text resolvers via function types
- Full `context.Context` support for cancellation

Here we'd like to thank go-git — we referenced a lot of go-git's code while building this VCS. We hope this merge implementation can contribute back to the community.

<details>
<summary><b>Architecture Overview</b></summary>

```
┌──────────────────────────────────────────────────────────────┐
│  Application Layer (pkg/zeta/)                               │
│  merge_tree.go — CLI entry, merge-base resolution, output    │
└──────────────────────────┬───────────────────────────────────┘
                           │
                           ▼
┌──────────────────────────────────────────────────────────────┐
│  Tree Merge Layer (pkg/zeta/odb/)                            │
│  merge.go       — three-way tree merge orchestration         │
│  merge_driver.go — text merge dispatch + charset restoration │
│  merge_text.go  — external merge tool integration            │
└──────────────────────────┬───────────────────────────────────┘
                           │
                           ▼
┌──────────────────────────────────────────────────────────────┐
│  Text Merge Layer (modules/diferenco/)                       │
│  merge.go    — diff3 algorithm, conflict detection & output  │
│  diferenco.go — diff algorithms (Histogram, Myers, ONP, ...) │
│  text.go     — charset detection, binary detection           │
│  sink.go     — line deduplication & indexing                  │
└──────────────────────────────────────────────────────────────┘
```

</details>

<details>
<summary><b>Text-Level Merge API</b></summary>

```go
// High-level: merge with default options (Histogram algorithm, merge style)
func DefaultMerge(ctx context.Context, o, a, b string, labelO, labelA, labelB string) (string, bool, error)

// Full control: specify algorithm and conflict style
func Merge(ctx context.Context, opts *MergeOptions) (string, bool, error)

// Quick check: does merging these texts produce conflicts?
func HasConflict(ctx context.Context, textO, textA, textB string) (bool, error)
```

Example:

```go
merged, hasConflict, err := diferenco.Merge(ctx, &diferenco.MergeOptions{
    TextO:  baseText,
    TextA:  oursText,
    TextB:  theirsText,
    LabelO: "base",
    LabelA: "ours",
    LabelB: "theirs",
    A:      diferenco.Histogram,
    Style:  diferenco.STYLE_ZEALOUS_DIFF3,
})
```

</details>

<details>
<summary><b>Tree-Level Merge Decision Logic</b></summary>

```
For each entry e in differences:
│
├─ Ancestor == Our   → accept Theirs (ours didn't change)
├─ Ancestor == Their → accept Ours (theirs didn't change)
├─ Our == Their      → accept either (both made same change)
└─ Otherwise (true conflict, enter mergeEntry)
    ├─ Both added same path
    │   ├─ binary/large → CONFLICT(binary), keep ours
    │   └─ text → three-way merge with empty base
    ├─ Both modified
    │   ├─ binary/large → CONFLICT(binary), keep ours
    │   └─ text → three-way merge with ancestor as base
    └─ One deletes, other modifies
        → CONFLICT(modify/delete), keep modified side
```

Conflict retention policy:

| Conflict Type | Retained in Result Tree |
|---------------|------------------------|
| Content | Merged text with conflict markers |
| Binary | Ours side |
| Modify/delete | The modified side |
| File/directory | File renamed to `path~branch` |
| Rename/rename | Both renames preserved |

</details>

<details>
<summary><b>Differences from Git</b></summary>

| Feature | Git (merge-ort) | HugeSCM |
|---------|----------------|---------|
| Rename detection | Fuzzy matching | Exact match only |
| Merge strategies | ours, theirs, octopus, subtree | Default only |
| Conflict styles | merge, diff3, zdiff3 | merge, diff3, zdiff3 |
| Diff algorithms | Myers, Histogram, Patience, Minimal | Histogram, Myers, ONP, Patience, Minimal, SuffixArray |
| Charset handling | None (assumes UTF-8) | Auto-detect and transcode |
| Hash algorithm | SHA-1 / SHA-256 | BLAKE3 |
| Implementation | C | Pure Go |

</details>

<details>
<summary><b>References</b></summary>

- [A Formal Investigation of Diff3](http://www.cis.upenn.edu/~bcpierce/papers/diff3-short.pdf) — Khanna, Kunal, Pierce (2007)
- [Merging with Diff3](https://blog.jcoglan.com/2017/05/08/merging-with-diff3/) — James Coglan
- [node-diff3](https://github.com/bhousel/node-diff3) — Original JavaScript implementation
- [epiclabs-io/diff3](https://github.com/epiclabs-io/diff3) — Go port we built upon

</details>

**Comment by akamensky** [comment]
I see the `.Repository` type now has `.Merge()` method (though for some unexplained reason it uses current reference instead of taking to/from references). Does that method now implement this feature?

**Comment by xeubie** [comment]
I implemented the three-way merge (and cherry-pick) in my own git library in case it helps anyone: https://github.com/xeubie/go-repo-mofo
